kotlin/app/src/main/java/com/example/chattutorial/ChannelActivity.kt

```diff
@@ -106,6 +63,59 @@ class ChannelActivity : AppCompatActivity(), PermissionRequestListener {
         // PermissionChecker.permissionCheck(getActivity(), this);
     }
 
+    private fun initViewModel(
+        channelType: String,
+        channelId: String
+    ) {
+
+        val viewModelFactory = ChannelViewModelFactory(application, channelType, channelId)
+        val viewModel = ViewModelProvider(this, viewModelFactory).get(ChannelViewModel::class.java)
+        val client = ChatClient.instance()
+        val channelController = client.channel(channelType, channelId)
+
+        viewModel.initialized.observe(this, Observer<Channel> { channel ->
+            // connect the view model
+            binding.viewModel = viewModel
+            val factory = MyMessageViewHolderFactory()
+            binding.messageList.setViewHolderFactory(factory)
+            binding.messageList.setViewModel(viewModel, this)
+            binding.messageInput.setViewModel(viewModel, this)
+            // binding.channelHeader.setViewModel(viewModel, this)
+            // If you are using own MessageInputView please comment this line.
+            binding.messageInput.setPermissionRequestListener(this)
+
+            val currentlyTyping = MutableLiveData<List<String>>(ArrayList())
+
+            channelController.events().subscribe {
+                val name = it.user?.name ?: ""
+                val typing = currentlyTyping.value ?: listOf()
+                val typingCopy: MutableList<String> = typing.toMutableList()
+                when (it) {
+                    is TypingStartEvent -> {
+                        if (typingCopy.contains(name).not()) {
+                            typingCopy.add(name)
+                        }
+                        currentlyTyping.postValue(typingCopy)
+                    }
+                    is TypingStopEvent -> {
+                        typingCopy.remove(name)
+                        currentlyTyping.postValue(typingCopy)
+                    }
+                }
+            }
+
+            val typingObserver = Observer<List<String>> { users ->
+                var typing = "nobody is typing"
+                if (users.isNotEmpty()) {
+                    typing = "typing: " + users.joinToString(", ")
+                }
+                binding.typing = typing
+            }
+            currentlyTyping.observe(this, typingObserver)
+        }
+        )
+    }
```

Comment:
This is a bit long for the tutorial, how can we shorten it?

Reply:
Well, there are just few new lines:
```
viewModel.initialized
```
and
```
channelController
```
Both are required. If `channelController` is removed, then client and filtering should be used, it will require probably the same few lines.